File: 99-recover-binary-search-tree/99-recover-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */

let first = null;
let second = null;
let temp = 0;
let prev = Number.MIN_VALUE;


function inorder(root){
    if(root == null) return;
    
    inorder(root.left);
    if(first == null && prev.val >root.val){
        first = prev;
    }
    
    if(first != null && prev.val >root.val){
        second = root;
    }
    
    prev = root;
    inorder(root.right);
    
}

var recoverTree = function(root) {
    first = null;
    second = null;
    temp = 0;
    prev = Number.MIN_VALUE;
    
    inorder(root);
    
    // console.log(first,second);
    temp = first.val;
    first.val = second.val;
    second.val = temp;
    
    // console.log(first,second,temp);
    
};

File: 99-recover-binary-search-tree/99-recover-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void recoverTree(TreeNode root) {
        //mrorris inorder traversal
        
        if(root == null) return;
        
        TreeNode prev = null;
        
        TreeNode first = null;
        TreeNode second = null;
        
        
        while(root != null){
            
            if(root.left == null){
                //print root
                if(prev != null && prev.val > root.val){
                    if(first == null){
                        first =  prev;
                        
                    }
                    
                    second = root;
                    
                }
                
                prev = root;
                
                root = root.right;
                
            }else{
                
                TreeNode cur = root.left;
                
                while(cur.right != null && cur.right != root) cur = cur.right;
                
                if(cur.right == null){
                    cur.right = root;
                    root = root.left;
                }else{
                    cur.right = null;
                    //print root
                    if(prev != null && prev.val > root.val){
                        if(first == null){
                            first =  prev;
                        
                        }
                    
                        second = root;
                    }
                     prev = root;
                    
                    root = root.right;
                }
                
            }
        }
        
        
        int temp = first.val;
        first.val = second.val;
        second.val = temp;
        
    }
}